beacon-chain/sync/block_batcher.go

```diff
@@ -0,0 +1,210 @@
+package sync
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	libp2pcore "github.com/libp2p/go-libp2p/core"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/db/filters"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+)
+
+// blockRangeBatcher encapsulates the logic for splitting up a block range request into fixed-size batches of
+// blocks that are retrieved from the database, ensured to be canonical, sequential and unique.
+// If a non-nil value for ticker is set, it will be used to pause between batches lookups, as a rate-limiter.
+type blockRangeBatcher struct {
+	start   primitives.Slot
+	end     primitives.Slot
+	size    uint64
+	db      db.NoHeadAccessDatabase
+	limiter *limiter
+	ticker  *time.Ticker
+
+	cf      *canonicalFilter
+	current *blockBatch
+}
+
+func newBlockRangeBatcher(rp rangeParams, bdb db.NoHeadAccessDatabase, limiter *limiter, canonical canonicalChecker, ticker *time.Ticker) *blockRangeBatcher {
+	cf := &canonicalFilter{canonical: canonical}
+	return &blockRangeBatcher{
+		start:   rp.start,
+		end:     rp.end,
+		size:    rp.size,
+		db:      bdb,
+		limiter: limiter,
+		ticker:  ticker,
+		cf:      cf,
+	}
+}
+
+func (bb *blockRangeBatcher) next(ctx context.Context, stream libp2pcore.Stream) (blockBatch, bool) {
+	var nb blockBatch
+	var more bool
+	// The result of each call to next() is saved in the `current` field.
+	// If current is not nil, current.next figures out the next batch based on the previous one.
+	// If current is nil, newBlockBatch is used to generate the first batch.
+	if bb.current != nil {
+		current := *bb.current
+		nb, more = current.next(bb.end, bb.size)
+	} else {
+		nb, more = newBlockBatch(bb.start, bb.end, bb.size)
+	}
+	// newBlockBatch and next() both return a boolean to indicate whether calling .next() will yield another batch
+	// (based on the whether we've gotten to the end slot yet). blockRangeBatcher.next does the same,
+	// and returns (zero value, false), to signal the end of the iteration.
+	if !more {
+		return blockBatch{}, false
+	}
+	if err := bb.limiter.validateRequest(stream, bb.size); err != nil {
+		return blockBatch{err: errors.Wrap(err, "throttled by rate limiter")}, false
+	}
+
+	// Wait for the ticker before doing anything expensive, unless this is the first batch.
+	if bb.ticker != nil && bb.current != nil {
+		<-bb.ticker.C
+	}
+	filter := filters.NewFilter().SetStartSlot(nb.start).SetEndSlot(nb.end)
+	blks, roots, err := bb.db.Blocks(ctx, filter)
+	if err != nil {
+		return blockBatch{err: errors.Wrap(err, "Could not retrieve blocks")}, false
+	}
+
+	// MMke slice with extra +1 capacity in case we want to grow it to also hold the genesis block.
```

Comment:
Where is the +1 capacity?

Reply:
oops, that comment is vestigial. I had a different approach to dealing with genesis but it made the code more complicated for little benefit. I'll :scissors: this comment.